crates/spartan-farmer/src/commands/dht/client.rs

```diff
@@ -0,0 +1,308 @@
+use super::*;
+use super::{core::create_node, eventloop::EventLoop};
+use libp2p::kad::QueryInfo;
+
+#[derive(Copy, Clone, Debug)]
+pub enum ClientType {
+    // Bootstrap node. It uses the following fields from `ClientConfig`:
+    // 1. `bootstrap_keys`: Private keys/private key location to create bootstrap node peerId.
+    // 2. `listen_addr`: Listening address for Bootstrap node.
+    Bootstrap,
+    // Normal node. It uses the following fields from `ClientConfig`:
+    // 1. `bootstrap_nodes`: Bootstrap nodes addresses that the normal node must connect to.
+    // For setting listening address, use client.start_listening.
+    Normal,
+}
+
+pub enum ClientEvent {
+    // Event for adding a listening address.
+    Listen {
+        addr: Multiaddr,
+        sender: oneshot::Sender<OneshotType>,
+    },
+    // Kademlia Random Walk for Peer Discovery. (GetClosestPeer)
+    #[allow(dead_code)]
+    RandomWalk {
+        key: Option<PeerId>,
+        sender: oneshot::Sender<QueryId>,
+    },
+    // List all known peers.
+    #[allow(dead_code)]
+    KnownPeers {
+        sender: oneshot::Sender<Vec<PeerId>>,
+    },
+    // Dial another peer.
+    #[allow(dead_code)]
+    Dial {
+        addr: Multiaddr,
+        peer: PeerId,
+        sender: oneshot::Sender<OneshotType>,
+    },
+    // Bootstrap during the initial connection to the DHT.
+    // NOTE: All the bootstrap nodes must already be connected to the swarm before we can start the
+    // bootstrap process.
+    Bootstrap {
+        sender: oneshot::Sender<QueryId>,
+    },
+    // Get all listening addresses.
+    #[allow(dead_code)]
+    Listeners {
+        sender: oneshot::Sender<Vec<Multiaddr>>,
+    },
+    // Read Kademlia Query Result.
+    #[allow(dead_code)]
+    QueryResult {
+        qid: QueryId,
+        sender: oneshot::Sender<String>,
+    },
+}
```

Comment:
It would be much better to remove this and have implementation in methods that you have already added..

There is a similar construction in `plot` module, but the reason it is there is because we have both read and write operations that should have different priorities, hence we have a custom loop there that prefers read operations over writes whenever possible. It doesn't seem like we have such requirements here, so it would be better to have async methods with necessary code in them. Much easier to read and maintain.

Reply:
I agree with you that this `ClientEvent` enum seems a bit unnecessary and is adding extra code that we do not need at this point. However, this is a common design pattern used in almost all major `libp2p` implementations in `ipfs`, `forest` and etc.

I will refer you to the following [comments](https://github.com/libp2p/rust-libp2p/pull/2186#issuecomment-895050359) and [discussion](https://github.com/libp2p/rust-libp2p/discussions/2024). 

That being said, we can remove it for the time being.

Reply:
I would very much prefer to use async mutex to access mutable state rather than even loop. Event loop still woks with channels that have locks inside of them one way or another.

Reply:
Can we keep this for now just to get bootstrap and peer_discovery working and then, replace it async mutex's later on?

Reply:
That is the point, I would very much prefer to not get the entire event loop into the code base in the first place.